// gluestack-ui/dist/config.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.config = void 0;
    const config = {
        repoUrl: 'https://github.com/gluestack/gluestack-ui.git',
        gluestackDir: '.gluestack/cache/gluestack-ui',
        componentsResourcePath: 'example/storybook-nativewind/src/core-components',
        hooksResourcePath: 'example/storybook-nativewind/src/core-components/hooks',
        dependencyConfigPath: 'example/storybook-nativewind/src/core-components/nativewind/dependencies.json',
        nativeWindRootPath: 'nativewind',
        expoProject: 'expo',
        nextJsProject: 'nextjs',
        reactNativeCLIProject: 'react-native-cli',
        tailwindConfigRootPath: 'example/storybook-nativewind/src/tailwind.config.js',
        writableComponentsPath: 'components/ui',
        branchName: 'main',
        alphaBranchName: 'feat/nativewind-4.1-support',
        style: 'nativewind',
        providerComponent: 'gluestack-ui-provider',
        nativewindUtilPattern: '@gluestack-ui/nativewind-utils/',
        gluestackUIPattern: '@gluestack-ui/',
        templatesDir: '../../../template',
        codeModesDir: '../../../template/codemods',
        packageManager: null,
    };
    exports.config = config;
});
